File: IntersectionofTwoLinkedLists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        la = self.length(headA)
        lb = self.length(headB)
        a = headA
        b = headB
        while la > lb:
            a = a.next
            la -= 1
        while lb > la:
            b = b.next
            lb -= 1
        while a != b:
            a = a.next
            b = b.next
        return a
        
    def length(self,head):
        res = 0
        n = head
        while n:
            n = n.next
            res += 1
        return res

File: intersectionOfTwoLinkedLists.kt
// refer to editorial solution III, also pay attention there could be case two list never intersect. 
// the solution brilliantly added a null phase during looping, so that when pointers finished 2 lists traversal, they will all concludes at a null
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun getIntersectionNode(headA:ListNode?, headB:ListNode?):ListNode? {
        var a = headA
        var b = headB
        while(a != b){
            a = if(a== null) headB else a.next
            b = if(b== null) headA else b.next
        }
        return a
    }
}
